src/track/trackmetadatataglib.cpp

```diff
@@ -1676,110 +1713,208 @@ Result readTrackMetadataAndCoverArtFromFile(TrackMetadata* pTrackMetadata, QImag
                 }
             }
         }
-    } else {
-        qWarning() << "Unsupported file type" << fileType;
+        break;
+    }
+    default:
+        qWarning()
+            << "Cannot read metadata from file"
+            << fileName
+            << "with unknown or unsupported file type"
+            << fileType;
+        return ERR;
     }
 
     qWarning() << "Failed to read track metadata from file" << fileName;
     return ERR;
 }
 
-Result writeTrackMetadataIntoFile(const TrackMetadata& trackMetadata, QString fileName) {
-    const QString fileType(getFileTypeFromFileName(fileName));
+// Encapsulates subtle differences between TagLib::File::save()
+// and variants of this function in derived subclasses.
+class TagSaver final {
```

Comment:
...but maybe I have another idea.

On 28.08.2016 14:17, Uwe Klotz wrote:
> That sounds like overkill for a simple (and ugly) workaround ;) I have to 
> think about it...
>
> On 28.08.2016 13:38, Daniel Schürmann wrote:
>>
>> In src/track/trackmetadatataglib.cpp 
>> <https://github.com/mixxxdj/mixxx/pull/997#discussion_r76530416>:
>>
>> >      }
>> >
>> >      qWarning() << "Failed to read track metadata from file" << fileName;
>> >      return ERR;
>> >  }
>> >
>> > -Result writeTrackMetadataIntoFile(const TrackMetadata& trackMetadata, QString fileName) {
>> > -    const QString fileType(getFileTypeFromFileName(fileName));
>> > +// Encapsulates subtle differences between TagLib::File::save()
>> > +// and variants of this function in derived subclasses.
>> > +class TagSaver final {
>>
>> Sorry, now it looks somehow half done, nothing that prevent merging, but 
>> it itches me to write: ;-)
>>
>> The original issue was: some code out of sight vs. code duplication. Now 
>> we have more code and still code out of sight.
>> How about turning the TagSaver glass into a abstract base class and 
>> Implement it for each file type?
>> This way we could meld don the whole case tree above.
>> Alternative we can move the whole case tree to one initOnce() function or 
>> to a initOnce() function for each type.
>>
>> This way we could get also around the unique pointer moving.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub 
>> <https://github.com/mixxxdj/mixxx/pull/997/files/a6f69573de0ced1d2bf83499679a94e0ab2925b9#r76530416>, 
>> or mute the thread 
>> <https://github.com/notifications/unsubscribe-auth/ABZ-aoGWYriavlCPmE86LR4lxBh8lAbdks5qkXM_gaJpZM4JustG>.
>>
>